Misc Chemistry Stuff + Major stimulant/antistun changes + Refactoring

Kierany9 opened the pull request:
### Code improvements
- Full refactor of how speed-boosting chemicals work code-wise. De-snowflakes a bunch of stuff and makes everything much easier to code in general, including some stuff that was flat out broken before. You no longer directly add these bitflags to the chemical, but rather adjust the speedboost var with defines.

### Bugfixes
- Fixed a bug where the speed increase from certain chemicals would be nullified by the IGNORESLOWDOWN bitflag.
- Fixed a bug where certain chemicals would not have bitflags added by them removed correctly, or were trying to remove the wrong bitflag(thanks to the refactor).
- Fixed Morphine pills instantly overdosing you and causing addiction by reducing the amount in them.

### Features
- **Morphine has been split into two chemicals, Sleep Toxin and Morphine.** The former is available in the sleepers and medvends at roundstart and can be used to sedate patients just as morphine was. The latter is available in upgraded sleepers and allows you to ignore all slowdowns, either induced by pain or by hardsuits. If you are near critical condition, it will put you to sleep. Morphine also halves the recharge rate of stun timers(see below). They can be made with 1 part choral hydrate 4 parts sugar and 1 part sleep toxin 1 part oil 1 part nitrogen respectively
- **Complete changing of the way anti-stun chemicals work.** Anti-stun chemicals no longer remove a certain amount of stun every tick, but rather allow you to completely or partially resist a stun after a short period of time. If you wait an extra five seconds after that and can avoid being stunned for a second time, you get another stun resist for free almost immediately. Being stunned will also cause the chemicals to metabolize twice as fast. This means that a single stun will do almost nothing but being stunned again in that timeframe will result in an easy stunlock and restraint instead of being able to shrug off so many stuns that their tasers/stunprods battery runs out. Detailed stats for each chemical are below.
- As a result of the above feature, **all previous nerfs to anti-stuns, including the wear-off effects, have been undone**
- **Adds Heroin.** Heroin, made with Morphine, Cyanide, Potassium and Epinephrine works as a far superior version of morphine. Heroin metabolizes at half the speed, grants unlimited stamina and has none of the negative side effects of morphine but unresisted stuns last twice as long. Despite being mildly toxic, it heals brute and burn damage as effectively as saline-glucose solution.
- Poppies now contain Morphine instead of Saline-Glucose solution, and at high potencies they contain Heroin.

Ephedrine: Removes 4 ticks of stun after being unstunned for 6 ticks. Non-toxic.
Methamphetamine: Removes 6 ticks of stun after being unstunned for 6 ticks. 0.6 tox damage per tick.
Crank: Removes 6 ticks of stun after being unstunned for 4 ticks. 0.15 tox damage per tick.
Bath Salts: Removes 12 ticks of stun after being unstunned for 2 ticks. 0.6 tox damage per tick.
Stimulants(Stimpack and Adrenals): Removes 4 ticks of stun after being unstunned for 4 ticks. Non-toxic.
Changeling Adrenaline Sacs: Removes 4 ticks of stun after being unstunned for 4 ticks. Non-toxic.
Unholy Water: Removes 6 ticks of stun after being unstunned for 4 ticks(Cultist only). 2 brain damage per tick.

### Minor stuff
- Atropine and Tricordiazine must be heated to 370k to create, this is to prevent accidental mixing.
- Charcoal, Mannitol and Mutadone pills now contain 15u instead of 50u, which was completely unnecesary and in the case of the former, annoying and detrimental.
- You now need to hack the medvend to get Diphenhydramine and Atropine from it.

Kierany9 commented:
DNM for now, no logs and there are a couple of balancing things that I need to iron out.

Kierany9 commented:
Alright, logs are here and stuff is ironed out, remove DNM and discuss.

JohnGinnane commented on code/game/gamemodes/changeling/powers/adrenaline.dm:
```diff
@@ -31,10 +31,19 @@
-	M.AdjustStunned(-1)
-	M.AdjustWeakened(-1)
-	M.adjustStaminaLoss(-1)
+	if(!(M.stunned || M.weakened || M.paralysis))
```
What's the purpose of this?

Kierany9 replied:
To check if they aren't stunned before increasing the timer, as well as resetting the metabolism.

JohnGinnane commented on code/modules/hydroponics/grown.dm:
```diff
@@ -149,7 +149,9 @@
 		reagents.add_reagent("nutriment", 1 + round((potency / 20), 1))
-		reagents.add_reagent("salglu_solution", 1 + round((potency / 20), 1))
+		reagents.add_reagent("morphine", 1 + round(potency / 10, 1))
+		if(potency > 60)
```
This means you'll get both morphine and heroin, correct?

Kierany9 replied:
Yes

JohnGinnane commented on code/modules/reagents/Chemistry-Holder.dm:
```diff
@@ -396,21 +396,21 @@ var/const/INJECT = 5 //injection
 /datum/reagents/proc/check_ignoreslow(mob/M)
 	if(istype(M, /mob))
-		if(M.reagents.has_reagent("morphine")||M.reagents.has_reagent("ephedrine"))
+		if(M.reagents.has_reagent("morphine")||M.reagents.has_reagent("heroin")||M.reagents.has_reagent("bath_salts"))
```
This way you can check if any reagent in body makes them fast

Kierany9 replied:
I could try. No idea why it was coded like this tbh.

JohnGinnane commented on code/modules/reagents/Chemistry-Reagents.dm:
```diff
@@ -25,6 +25,7 @@
 	var/addiction_threshold = 0
 	var/addiction_stage = 0
 	var/overdosed = 0 // You fucked up and this is now triggering its overdose effects, purge that shit quick.
+	var/stun_timer = 0 // How many ticks since you last resisted a stun, tracked by stun-resisting reagents.
```
Could also add the threshold to when stuns can be resisted again, E.g morphine's would be stun_threshold=4

Kierany9 replied:
Eh sure.

JohnGinnane replied:
Sorry, was looking at changling adrenaline's stun timing..

JohnGinnane commented on code/modules/reagents/Chemistry-Reagents/Drug-Reagents.dm:
```diff
@@ -56,56 +57,55 @@
+	metabolization_rate = 0.75* REAGENTS_METABOLISM
+	stun_threshold = 4
 
 /datum/reagent/drug/crank/on_mob_life(mob/living/M)
```
It looks as though this and crank have very very similar on_mob_life definitions. Is there anything we can do to make it less repetitive?

Carbonhell commented:
is this pr ready to be codechecked?

McBawbaggings commented on code/modules/reagents/Chemistry-Reagents/Drug-Reagents.dm:
```diff
@@ -355,10 +352,95 @@
+
+/datum/reagent/drug/heroin/on_mob_life(mob/living/M)
+	M.setStaminaLoss(0)
+	if(M.stunned || M.weakened || M.paralysis)
```
Is there any point for checking M.paralysis?
Since you're not extending the paralysis, all you're doing is adding an extra 0.5 per tick to their stuns which they'll be stunned anyway?

McBawbaggings commented on code/modules/reagents/Chemistry-Reagents/Drug-Reagents.dm:
```diff
@@ -355,10 +352,95 @@
+			M.drop_item()
+		M.Dizzy(2)
+		M.Jitter(2)
+	if(prob(10))
```
This like a 1% or less chance to end up with >7 sleeping.
It'd be nice if it checked you weren't already asleep though.

McBawbaggings commented on code/modules/reagents/Chemistry-Reagents/Medicine-Reagents.dm:
```diff
@@ -538,28 +548,61 @@
 	..()
 	return
 
+/datum/reagent/medicine/sleeptoxin
```
Nit picking but wouldn't this be better as a toxin with toxpower being 0?
Chloral Hydrate does the same thing.

McBawbaggings commented on code/game/gamemodes/changeling/powers/adrenaline.dm:
```diff
@@ -29,12 +29,22 @@
-	M.adjustStaminaLoss(-1)
+	if(!(M.stunned || M.weakened || M.paralysis))
+		stun_timer++
+		metabolization_rate = REAGENTS_METABOLISM
```
Could you not use initial(metabolization_rate) is your multiplier?
So if someone wants to change the metabolizationrate, they only need to change the value at the definitions of the reagent?

McBawbaggings commented on code/game/gamemodes/changeling/powers/adrenaline.dm:
```diff
@@ -29,12 +29,22 @@
+		metabolization_rate = REAGENTS_METABOLISM
+		M.adjustStaminaLoss(-1)
+	else
+		metabolization_rate = 2 * REAGENTS_METABOLISM
```
Could you not use initial(metabolization_rate) is your multiplier?
So if someone wants to change the metabolizationrate, they only need to change the value at the definitions of the reagent?

McBawbaggings commented on code/modules/reagents/Chemistry-Reagents/Drug-Reagents.dm:
```diff
@@ -56,56 +57,55 @@
+		M << "<span class='notice'>[high_message]</span>"
+	if(!(M.stunned || M.weakened || M.paralysis))
+		stun_timer++
+		metabolization_rate = 0.75* REAGENTS_METABOLISM
```
Could you not use initial(metabolization_rate) is your multiplier?
So if someone wants to change the metabolizationrate, they only need to change the value at the definitions of the reagent?

McBawbaggings commented on code/modules/reagents/Chemistry-Reagents/Drug-Reagents.dm:
```diff
@@ -56,56 +57,55 @@
+		stun_timer++
+		metabolization_rate = 0.75* REAGENTS_METABOLISM
+	else
+		metabolization_rate = 1.5 * REAGENTS_METABOLISM
```
Could you not use initial(metabolization_rate) is your multiplier?
So if someone wants to change the metabolizationrate, they only need to change the value at the definitions of the reagent?

McBawbaggings commented on code/modules/reagents/Chemistry-Reagents/Drug-Reagents.dm:
```diff
@@ -168,40 +168,34 @@
+		M << "<span class='notice'>[high_message]</span>"
+	if(!(M.stunned || M.weakened || M.paralysis))
+		stun_timer++
+		metabolization_rate = 0.75* REAGENTS_METABOLISM
```
Could you not use initial(metabolization_rate) is your multiplier?
So if someone wants to change the metabolizationrate, they only need to change the value at the definitions of the reagent?

McBawbaggings commented on code/modules/reagents/Chemistry-Reagents/Drug-Reagents.dm:
```diff
@@ -168,40 +168,34 @@
+		stun_timer++
+		metabolization_rate = 0.75* REAGENTS_METABOLISM
+	else
+		metabolization_rate = 1.5 * REAGENTS_METABOLISM
```
Could you not use initial(metabolization_rate) is your multiplier?
So if someone wants to change the metabolizationrate, they only need to change the value at the definitions of the reagent?

McBawbaggings commented on code/modules/reagents/Chemistry-Reagents/Drug-Reagents.dm:
```diff
@@ -260,56 +255,58 @@
 	description = "Makes you nearly impervious to stuns and grants a stamina regeneration buff, but you will be a nearly uncontrollable tramp-bearded raving lunatic."
 	reagent_state = LIQUID
 	color = "#60A584" // rgb: 96, 165, 132
+	metabolization_rate = 0.5 * REAGENTS_METABOLISM
```
Could you not use initial(metabolization_rate) is your multiplier?
So if someone wants to change the metabolizationrate, they only need to change the value at the definitions of the reagent?

McBawbaggings commented on code/modules/reagents/Chemistry-Reagents/Drug-Reagents.dm:
```diff
@@ -260,56 +255,58 @@
-	if(current_cycle >= 5)
+	if(!(M.stunned || M.weakened || M.paralysis))
+		stun_timer++
+		metabolization_rate = 0.5* REAGENTS_METABOLISM
```
Could you not use initial(metabolization_rate) is your multiplier?
So if someone wants to change the metabolizationrate, they only need to change the value at the definitions of the reagent?

McBawbaggings commented on code/modules/reagents/Chemistry-Reagents/Drug-Reagents.dm:
```diff
@@ -260,56 +255,58 @@
+		metabolization_rate = 0.5* REAGENTS_METABOLISM
+		M.adjustStaminaLoss(-4)
+	else
+		metabolization_rate = REAGENTS_METABOLISM
```
Could you not use initial(metabolization_rate) is your multiplier?
So if someone wants to change the metabolizationrate, they only need to change the value at the definitions of the reagent?

McBawbaggings commented on code/modules/reagents/Chemistry-Reagents/Medicine-Reagents.dm:
```diff
@@ -468,13 +468,23 @@
-	M.adjustStaminaLoss(-0.5*REM)
+	if(!(M.stunned || M.weakened || M.paralysis))
+		stun_timer++
+		metabolization_rate = 0.5* REAGENTS_METABOLISM
```
Could you not use initial(metabolization_rate) is your multiplier?
So if someone wants to change the metabolizationrate, they only need to change the value at the definitions of the reagent?

McBawbaggings commented on code/modules/reagents/Chemistry-Reagents/Medicine-Reagents.dm:
```diff
@@ -468,13 +468,23 @@
+		metabolization_rate = 0.5* REAGENTS_METABOLISM
+		M.adjustStaminaLoss(-1)
+	else
+		metabolization_rate = 2 * REAGENTS_METABOLISM
```
Could you not use initial(metabolization_rate) is your multiplier?
So if someone wants to change the metabolizationrate, they only need to change the value at the definitions of the reagent?

McBawbaggings commented on code/modules/reagents/Chemistry-Reagents/Medicine-Reagents.dm:
```diff
@@ -806,19 +849,29 @@
-	M.adjustStaminaLoss(-5*REM)
+	if(!(M.stunned || M.weakened || M.paralysis))
+		stun_timer++
+		metabolization_rate = 0.5* REAGENTS_METABOLISM
```
Could you not use initial(metabolization_rate) is your multiplier?
So if someone wants to change the metabolizationrate, they only need to change the value at the definitions of the reagent?

McBawbaggings commented on code/modules/reagents/Chemistry-Reagents/Medicine-Reagents.dm:
```diff
@@ -806,19 +849,29 @@
+		metabolization_rate = 0.5* REAGENTS_METABOLISM
+		M.adjustStaminaLoss(-3)
+	else
+		metabolization_rate = REAGENTS_METABOLISM
```
Could you not use initial(metabolization_rate) is your multiplier?
So if someone wants to change the metabolizationrate, they only need to change the value at the definitions of the reagent?

Kierany9 commented:
Yes, it's ready

Kierany9 commented:
Charcoal, Mutadone and Mannitol pills now contain 15u each instead of 50u. I probably don't need to explain why but 50u antitox will purge chems from your body for about five minutes and Mutadone/Mannitol take 1u or less to work, having 50 for the sake of it is just absurd.

Morphine pills now contain 10u instead of 30u so you don't fucking overdose and and get addicted in a single pill.

JohnGinnane commented:
what about using a pill on a reagent container (beakers, beer bottles, glasses, etc.) will put the reagents into the container, effectively dissolving the pill. This way you can spike stuff

Kierany9 commented:
Has been a feature for ages

McBawbaggings commented:
I think these are good changes. +1

Badmex commented:
Yeah +1/3 I like it

Kierany9 commented:
After some actual testing and talking with a few people, I've made a few pretty big changes. Discuss.

- **You can now resist two stuns instead of just one**, provided you wait three extra ticks and spend at least one tick unstunned after resisting your first stun.
- You now get a message when the chems are ready to resist.
- Moved the copypaste stun resist code on every single chem into its own proc.
- Removed the staminaloss while stunned because it doesn't really come into play and because it's more copypaste code.
- Stun resistance for each anistun now stacks. This only really comes into play with either weaker antistuns or very long stuns, as all of them can instantly resist a taser bolt and half of them resist a stun baton.

(also still waiting for a codecheck)

JohnGinnane commented:
Is this still WIP?

Kierany9 commented:
Not really, no

Nebyab commented:
sure

Carbonhell commented on code/modules/reagents/Chemistry-Reagents.dm:
```diff
@@ -2,6 +2,11 @@
 #define GAS 3
 
+#define NORMAL 0
+#define FAST 1
```
if you wanna use those defines elsewhere you better move them in some define dm file

Carbonhell commented on code/modules/reagents/Chemistry-Holder.dm:
```diff
@@ -389,31 +390,20 @@ var/const/INJECT = 5 //injection
 
-/datum/reagents/proc/check_ignoreslow(mob/M)
+/datum/reagents/proc/check_speedboost(mob/M)
 	if(istype(M, /mob))
```
could just say if(istype(M)) since it's typecasted

Carbonhell commented on code/modules/reagents/Chemistry-Reagents.dm:
```diff
@@ -113,3 +122,21 @@
 
 	return result
+
+datum/reagent/proc/stun_resist_act(mob/living/M)
```
forgot the starting slash

Carbonhell commented:
ladies and gentlemen, merging this

Kierany9 commented:
>75 commits
please no more refactoring
